python/ml4ir/base/features/feature_fns/categorical.py

```diff
@@ -0,0 +1,260 @@
+import tensorflow as tf
+from tensorflow.keras import layers
+from tensorflow import feature_column
+from tensorflow import lookup
+
+import copy
+
+from ml4ir.base.io import file_io
+
+
+def categorical_embedding_with_hash_buckets(feature_tensor, feature_info):
+    """
+    Convert a string feature tensor into a categorical embedding.
+    Works by first converting the string into num_hash_buckets buckets
+    each of size hash_bucket_size, then converting each hash bucket into
+    a categorical embdding of dimension embedding_size. Finally, these embeddings
+    are combined either through mean, sum or concat operations to generate the final
+    embedding based on the feature_info.
+
+    Args:
+        feature_tensor: String feature tensor
+        feature_info: Dictionary representing the configuration parameters for the specific feature from the FeatureConfig
+
+    Returns:
+        categorical embedding for the input feature_tensor
+
+    Args under feature_layer_info:
+        num_hash_buckets: int; number of different hash buckets to convert the input string into
+        hash_bucket_size: int; the size of each hash bucket
+        embedding_size: int; dimension size of the categorical embedding
+        merge_mode: str; can be one of "mean", "sum", "concat" representing the mode of combining embeddings from each categorical embedding
+    """
+
+    # Numeric input features
+    if feature_info["dtype"] in (tf.float32, tf.int64):
+        raise NotImplementedError
+
+    # String input features
+    elif feature_info["dtype"] in (tf.string,):
```

Comment:
The code exits `if feature_info["dtype"] not in (tf.float32, tf.int64)` e.g., `if not feature_info["dtype"] ==tf.string; raise ..`
Given the above, the `elif` statement can be removed and the code below visually simplified removing indentation.

Reply:
That's true. I should probably just remove the `tf.float32` and `tf.int64` from this method as well. Not really needed.